src/main/java/org/testng/internal/BaseTestMethod.java

```diff
@@ -428,7 +428,11 @@ public boolean equals(Object obj) {
    */
   @Override
   public int hashCode() {
-    return m_method.hashCode();
+    int hash = m_method.hashCode();
+    if (m_instance != null) {
+      hash = hash * 31 + m_instance.hashCode();
+    }

```

Comment:
`BaseTestMethod#equals` uses `identity` comparison like `getInstance() == other.getInstance()`, so the `hashCode` implementation should use identity hashCode rather than user-provided implementation.

In other words, this should be `System.identityHashCode(m_instance)`

Other than that, the change is a clear improvement with little-to-no risk, so it is safe to merge if `m_instance.hashCode()` is replaced with `identityHashCode`

Reply:
I'm confused. How is this a performance improvement? It does more work than the previous implementation.

Reply:
Extra work is negligible, and the new implementation produces better hash codes, so there are less chances for collision in

```
   at java.util.HashSet.add
   at org.testng.internal.DynamicGraph.setStatus 
   at org.testng.internal.thread.graph.GraphThreadPoolExecutor.setStatus
```